File: ros2_ws/src/rcdt_utilities/src_py/manipulate_pose.py
#!/usr/bin/env python3

# SPDX-FileCopyrightText: Alliander N. V.
#
# SPDX-License-Identifier: Apache-2.0

from copy import copy

import rclpy
from geometry_msgs.msg import Point, PoseStamped, Transform, TransformStamped
from rcdt_messages.srv import ExpressPoseInOtherFrame, TransformPose
from rcdt_utilities.launch_utils import spin_node
from rclpy import logging, time
from rclpy.node import Node
from tf2_geometry_msgs import do_transform_pose_stamped
from tf2_ros import TransformException
from tf2_ros.buffer import Buffer
from tf2_ros.transform_listener import TransformListener

ros_logger = logging.get_logger(__name__)


class ManipulatePose(Node):
    def __init__(self):
        super().__init__("manipulate_pose")

        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)
        self.create_service(
            ExpressPoseInOtherFrame,
            "/express_pose_in_other_frame",
            self.express_pose_in_other_frame,
        )
        self.create_service(
            TransformPose,
            "/transform_pose",
            self.transform_pose,
        )

    def express_pose_in_other_frame(
        self,
        request: ExpressPoseInOtherFrame.Request,
        response: ExpressPoseInOtherFrame.Response,
    ) -> ExpressPoseInOtherFrame.Response:
        source_frame = request.pose.header.frame_id
        target_frame = request.target_frame

        try:
            transform = self.tf_buffer.lookup_transform(
                target_frame, source_frame, time.Time()
            )
        except TransformException as ex:
            self.get_logger().error(
                f"Could not transform {target_frame} to {source_frame}: {ex}"
            )
            response.success = False
            return response
        response.pose = do_transform_pose_stamped(request.pose, transform)
        response.success = True
        return response

    def transform_pose(
        self, request: TransformPose.Request, response: TransformPose.Response
    ) -> TransformPose.Response:
        response.pose = apply_transform(request.pose, request.transform)
        response.success = True
        return response

    def transform_pose_relative(
        self, request: TransformPose.Request, response: TransformPose.Response
    ) -> TransformPose.Response:
        response.pose = apply_transform_relative(request.pose, request.transform)
        response.success = True
        return response


def apply_transform(pose: PoseStamped, transform: Transform) -> PoseStamped:
    transform_stamped = TransformStamped()
    transform_stamped.transform = transform
    transform_stamped.header = pose.header
    return do_transform_pose_stamped(pose, transform_stamped)


def apply_transform_relative(pose: PoseStamped, transform: Transform) -> PoseStamped:
    start_position = copy(pose.pose.position)
    pose.pose.position = Point()
    pose = apply_transform(pose, transform)
    pose.pose.position.x += start_position.x
    pose.pose.position.y += start_position.y
    pose.pose.position.z += start_position.z
    return pose


def main(args: str = None) -> None:
    rclpy.init(args=args)
    node = ManipulatePose()
    spin_node(node)


if __name__ == "__main__":
    main()

File: ros2_ws/src/rcdt_utilities/src/manipulate_pose.cpp
// # SPDX-FileCopyrightText: Alliander N. V.
//
// # SPDX-License-Identifier: Apache-2.0

#include "rcdt_utilities/manipulate_pose.hpp"

PoseManipulator::PoseManipulator() : rclcpp::Node("pose_manipulator") {
  tf_buffer_ = std::make_unique<tf2_ros::Buffer>(this->get_clock());
  tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);

  express_pose_in_other_frame_service_ = this->create_service<
      rcdt_interfaces::srv::ExpressPoseInOtherFrame>(
      "~/express_pose_in_other_frame",
      [this](
          std::shared_ptr<rcdt_interfaces::srv::ExpressPoseInOtherFrame::Request>
              req,
          std::shared_ptr<rcdt_interfaces::srv::ExpressPoseInOtherFrame::Response>
              resp) { this->expressPoseInOtherFrame(req, resp); });

  transform_pose_service_ =
      this->create_service<rcdt_interfaces::srv::TransformPose>(
          "~/transform_pose",
          [this](
              std::shared_ptr<rcdt_interfaces::srv::TransformPose::Request> req,
              std::shared_ptr<rcdt_interfaces::srv::TransformPose::Response>
                  resp) { this->transformPose(req, resp); });

  transform_pose_relative_service_ =
      this->create_service<rcdt_interfaces::srv::TransformPose>(
          "~/transform_pose_relative",
          [this](
              std::shared_ptr<rcdt_interfaces::srv::TransformPose::Request> req,
              std::shared_ptr<rcdt_interfaces::srv::TransformPose::Response>
                  resp) { this->transformPoseRelative(req, resp); });
}

void PoseManipulator::expressPoseInOtherFrame(
    const std::shared_ptr<rcdt_interfaces::srv::ExpressPoseInOtherFrame::Request>
        req,
    std::shared_ptr<rcdt_interfaces::srv::ExpressPoseInOtherFrame::Response>
        resp) {
  std::string source_frame = req->pose.header.frame_id;
  std::string target_frame = req->target_frame;

  geometry_msgs::msg::TransformStamped tf;
  try {
    tf = tf_buffer_->lookupTransform(target_frame, source_frame,
                                     rclcpp::Time(0));
  } catch (const tf2::TransformException& e) {
    RCLCPP_WARN(this->get_logger(),
                "Could not transform from frame %s to frame %s: %s",
                source_frame.c_str(), target_frame.c_str(), e.what());
    resp->success = false;
    return;
  }

  if (req->pose.header.frame_id != tf.child_frame_id) {
    RCLCPP_ERROR(this->get_logger(),
                 "Pose frame ID (%s) and Transform child frame ID (%s) are not "
                 "the same.",
                 req->pose.header.frame_id.c_str(), tf.child_frame_id.c_str());
  }

  try {
    tf2::doTransform(req->pose, resp->pose, tf);
  } catch (const tf2::TransformException& e) {
    RCLCPP_ERROR(this->get_logger(),
                 "Could not transform pose to frame %s: %s.",
                 target_frame.c_str(), e.what());
  }
  resp->success = true;
}

void PoseManipulator::transformPose(
    const std::shared_ptr<rcdt_interfaces::srv::TransformPose::Request> req,
    std::shared_ptr<rcdt_interfaces::srv::TransformPose::Response> resp) {
  geometry_msgs::msg::TransformStamped tf_stamped;
  tf_stamped.transform = req->transform;
  tf2::doTransform(req->pose, resp->pose, tf_stamped);

  resp->success = true;
}

void PoseManipulator::transformPoseRelative(
    const std::shared_ptr<rcdt_interfaces::srv::TransformPose::Request> req,
    std::shared_ptr<rcdt_interfaces::srv::TransformPose::Response> resp) {
  geometry_msgs::msg::TransformStamped tf_stamped;
  tf_stamped.transform = req->transform;
  tf2::doTransform(req->pose, resp->pose, tf_stamped);

  resp->pose.pose.position.x += req->pose.pose.position.x;
  resp->pose.pose.position.y += req->pose.pose.position.y;
  resp->pose.pose.position.z += req->pose.pose.position.z;

  resp->success = true;
}

geometry_msgs::msg::PoseStamped PoseManipulator::doTransform(
    const geometry_msgs::msg::PoseStamped pose_in,
    const geometry_msgs::msg::Transform tf) {
  geometry_msgs::msg::PoseStamped pose_out;
  return pose_out;
}
